Community/GenericPresence/automation/jsr223/python/community/genericpresence/presence.py

```diff
@@ -0,0 +1,159 @@
+                        "gRichPresent": ("vRichPresent",    3),
+                        "gJennPresent": ("vJennPresent",    2) }
+
+Examples
+========
+.. code-block::
+    # Example Items:
+    // All people
+    Group:Switch:OR(ON,OFF) gPresent
+    Switch vPresent
+    Switch aPresenceOverride (gPresent)
+    // Rich
+    Group:Switch:OR(ON,OFF) gRichPresent (gPresent)
+    Switch vRichPhone (gRichPresent)
+    Switch vRichPhone_Manticore_Net (gRichPresent)
+    // Jenn
+    Group:Switch:OR(ON,OFF) gJennPresent (gPresent)
+    Switch vJennPhone (gJennPresent)
+    Switch vJennPhone_Manticore_Net (gJennPresent)
+
+Notice how the there are two sensors for each person which aggregate to a
+Group. The two person Groups are made members of gPresent. There is an override
+Switch to manually set present to ON.
+"""
+from core.rules import rule
+from core.triggers import when
+from configuration import presence_config
+from core.actions import ScriptExecution
+from org.joda.time import DateTime
+
+presence_timers = {}
+
+def trigger_generator(presence_config):
+    """
+    Called to generate triggers for all of the defined presence sensors imported
+    from configuration.
+
+    Arguments:
+        - presence_config: A dict whose keys are the names of the sensor Items
+        that are managed by this Rule.
+    """
+    def generated_triggers(function):
+        for sensor in list(presence_config):
+            when("Item {} changed".format(sensor))(function)
+        return function
+    return generated_triggers
+
+def away(log, events, proxy_name, flap_time):
+    """
+    Called when a person has been detected away for more than the flapping time.
+    Log out the away state and command the Item to OFF.
+
+    Arguments:
+        - log: Logger passed in from the presence Rule.
+        - events: Access to the events Object so we can call sendCommand.
+        - proxy_name: Name of the proxy Item that we want to set to away.
+        - flap_time: Number of minutes we wait before marking the person away.
+    """
+    log.info("{} has been away for {} minutes, setting to away."
+            .format(proxy_name,flap_time))
+    events.sendCommand(proxy_name, "OFF")
+
+@rule("Presence",
+      description=("Aggregates presence sensor readings and implements "
+                   "anti-flapping for away."),
+      tags=["presence"])
+@trigger_generator(presence_config)
+def presence(event):
+    """
+    Rule that implements presence detection based on the state of sensors. The
+    Rule assumes there is a proxy Item associated with each sensor which
+    represents the presence for that sensor. There is flapping logic which
+    requires the user to be away for some minutes before the proxy will be set
+    to OFF.
+    """
+
+    # Get the proxy Item associated with the sensor.
+    try:
+        sensor_name = event.itemName
+        proxy_name = presence_config[sensor_name][0]
+        flap_time = presence_config[sensor_name][1]
```

Comment:
This would be a safer way to do this. You want to catch errors in the above 3 lines only, but you might catch one in the rest of it.

```python
    try:
        sensor_name = event.itemName
        proxy_name = presence_config[sensor_name][0]
        flap_time = presence_config[sensor_name][1]
    except ValueError:
        presence.log.error("{} is not in presence_sensors!"
                           .format(event.itemName))
    except IndexError:
        presence.log.error("presence_items does not have the same number of "
                           "elements as presence_sensors, check your "
                           "configuration!")
    else:
        if items[sensor_name] == items[proxy_name]:
            if (proxy_name in presence_timers.keys()
                    and not presence_timers[proxy_name].hasTerminated()):
                presence.log.debug("{} came home, cancelling the flapping "
                                   "timer.".format(proxy_name))
                presence_timers[proxy_name].cancel()
                del presence_timers[proxy_name]
            return

        # Item and sensor are different, set a flapping timer if the sensor went
        # to OFF.
        if event.itemState == OFF:
            presence.log.debug("Setting the flapping timer for {}"
                               .format(proxy_name))
            presence_timers = ScriptExecution.createTimer(
                DateTime.now().plusMinutes(presence_flap_time),
                lambda: away(presence.log, events, proxy_name, presence_timers))
        elif event.itemState == ON:
            presence.log.debug("{} came home.".format(proxy_name))
            events.sendCommand(proxy_name, "ON")
```

Reply:
I actually wrote it like that to begin with but I kept getting syntax and runtime errors on the else:. I was supposed to come back and figure out why and I guess I forgot. It seems to be working correctly now. This probably explains why sometimes I would get the error that presence_timers doesn't have a getKeys() method in the unload function. 

This is taking some getting used to. With Rules DSL I know that if VSCode shows no errors the Rule will at least run. With Python I have to run it just to discover some common typos. This is going to be hard a frustrating for users migrating from Rules DSL.

Reply:
The problem is the environment. We can't fake the OH script extension context to test the rules in without tremendous effort. At least rules are loaded when the files are modified so you don't have to restart OH every time. Installing pylint can help with obvious stuff, but is also unaware of the OH context